aptos-move/framework/aptos-framework/sources/object.move

```diff
@@ -0,0 +1,433 @@
+/// * Secure and guaranteed to not collide with existing accounts.
+///   * Each address is generated in a collision-resistant fashion guaranteeing legitimate
+///     accounts cannot be created within the same address.
+///   * Two objects cannot be co-located due to an address being able to only store a single Object
+///
+/// TODO:
+/// * There is no means to borrow an object or a reference to an object. We are exploring how to
+///   make it so that a reference to a global object can be returned from a function.
+module aptos_framework::object {
+    use std::bcs;
+    use std::error;
+    use std::hash;
+    use std::signer;
+    use std::vector;
+
+    use aptos_framework::account;
+    use aptos_framework::event;
+    use aptos_framework::from_bcs;
+    use aptos_framework::guid;
+
+    /// An object already exists at this address
+    const EOBJECT_EXISTS: u64 = 0;
+    /// An object of that type already exists at this address
+    const EOBJECT_TYPE_EXISTS: u64 = 1;
+    /// An object does not exist at this address
+    const EOBJECT_DOES_NOT_EXIST: u64 = 2;
+    /// The object does not have ungated transfers enabled
+    const ENO_UNGATED_TRANSFERS: u64 = 3;
+    /// The caller does not have ownership permissions
+    const ENOT_OBJECT_OWNER: u64 = 4;
+    /// The object does not allow for deletion
+    const ECANNOT_DELETE: u64 = 5;
+
+    /// Scheme identifier used to generate an object's address. This serves as domain separation to
+    /// prevent existing authentication key and resource account derivation to produce an object
+    /// address.
+    const OBJECT_ADDRESS_SCHEME: u8 = 0xFE;
+
+    #[resource_group_member(group = aptos_framework::object::ObjectGroup)]
+    struct Object has key {
+        guid_creation_num: u64,
+        owner: address,
+        allow_ungated_transfer: bool,
+        transfer_events: event::EventHandle<TransferEvent>,
+    }
+
+    #[resource_group(scope = global)]
+    struct ObjectGroup { }
+
+    /// Type safe way of designate an object as at this address
+    struct ObjectId has copy, drop, store {
+        inner: address,
+    }
+
+    /// This is a one time ability given to the creator to configure the object as necessary
+    struct CreatorRef has drop {
+        self: ObjectId,
+        can_delete: bool,
+    }
+
+    /// The owner of this ability can delete an object
+    struct DeleteRef has drop, store {
+        self: ObjectId,
+    }
+
+    /// Access to this ref allows for creation of the signer to add new resources
+    struct ExtendRef has drop, store {
+        self: ObjectId,
+    }
+
+    /// Access to this ref allows for owner to transfer the resources between two addresses
+    struct TransferRef has drop, store {
+        self: ObjectId,
+    }
+
+    /// Access to this ref allows for owner to transfer the resources between two addresses
+    struct LinearTransferRef has drop {
+        self: ObjectId,
+    }
+
+    struct TransferEvent has drop, store {
```

Comment:
Should we add events for when refs are created? This can be important for users of a smart contract to know when refs are created that can affect their trust assumptions of the objects that smart contract creates.

Reply:
Hmm....